lib/recursive-methods.rb

```diff
@@ -1,49 +1,106 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError, "must be a number greater than 0." if n < 0
+
+  return 1 if n == 1
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(1)
+
+def reverse(string, rev = "", first = 0)
+  if rev.length == string.length
+    return rev
+  end
+  rev = string[first] + rev
```

Comment:
String concatenation creates a new array and  copies all the individual elements over and so is O(n) by itself.